nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/actions/CSVExportAction.java

```diff
@@ -0,0 +1,309 @@
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.Serializable;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.nuxeo.ecm.core.api.Blob;
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.ecm.core.api.impl.blob.FileBlob;
+import org.nuxeo.ecm.core.bulk.BulkCodecs;
+import org.nuxeo.ecm.core.bulk.BulkCounter;
+import org.nuxeo.ecm.core.bulk.BulkRecords;
+import org.nuxeo.ecm.core.bulk.BulkStatus;
+import org.nuxeo.ecm.core.io.download.DownloadService;
+import org.nuxeo.ecm.core.transientstore.api.TransientStore;
+import org.nuxeo.ecm.core.transientstore.api.TransientStoreService;
+import org.nuxeo.ecm.core.utils.BlobUtils;
+import org.nuxeo.lib.stream.computation.AbstractComputation;
+import org.nuxeo.lib.stream.computation.Computation;
+import org.nuxeo.lib.stream.computation.ComputationContext;
+import org.nuxeo.lib.stream.computation.Record;
+import org.nuxeo.lib.stream.computation.Topology;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.kv.KeyValueService;
+import org.nuxeo.runtime.kv.KeyValueStore;
+
+/**
+ * @since 10.3
+ */
+public class CSVExportAction extends AbstractBulkAction {
+
+    public static final String ACTION_NAME = "csvExport";
+
+    public CSVExportAction() {
+        super(ACTION_NAME);
+    }
+
+    @Override
+    public Topology getTopology(Map<String, String> options) {
+        int size = getOptionAsInteger(options, BATCH_SIZE_OPT, DEFAULT_BATCH_SIZE);
+        int timer = getOptionAsInteger(options, BATCH_THRESHOLD_MS_OPT, DEFAULT_BATCH_THRESHOLD_MS);
+        return Topology.builder()
+                       .addComputation(() -> new DocumentCSVProjection(size, timer),
+                               Arrays.asList("i1:" + ACTION_NAME, "o1:" + FileMaterialization.NAME))
+                       .addComputation(FileMaterialization::new,
+                               Arrays.asList("i1:" + FileMaterialization.NAME, "o1:" + SortBlobLines.NAME))
+                       .addComputation(SortBlobLines::new,
+                               Arrays.asList("i1:" + SortBlobLines.NAME, "o1:" + ZipBlob.NAME))
+                       .addComputation(ZipBlob::new,
+                               Arrays.asList("i1:" + ZipBlob.NAME, "o1:" + ExposeBlobInDownload.NAME))
+                       .addComputation(ExposeBlobInDownload::new,
+                               Arrays.asList("i1:" + ExposeBlobInDownload.NAME, "o1:" + COUNTER_ACTION_NAME))
+                       .build();
+    }
+
+    @Override
+    protected Computation createComputation(int size, int timer) {
+        throw new UnsupportedOperationException();
+    }
```

Comment:
I think this method is not relevant anymore, could we rather have an abstract `buildTopology` method as an action can be composed of 1 or multiple computations ?